### PR TITLE
[feat] Rewrite the input projection + add several init options

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Loosely linked to #219 and #264

- rewrite the input projection to make it a little cleaner and easier to follow
  - nn.Linear instead of reinventing the linear projection wheel
  - explicit setting for self attention
  - move the weight init settings out of it (this was the only part which exposed something specific)
  - renam in_proj_container into input_projection, it feels a little easier to understand ?
  
  
- add a dedicated weight_init file (purely optional, opt in basis)
  - handle deepnorm from there
  - handle a couple of popular options for weight inits in general

Many TODOS left, comments welcome
- [x] Rewrite input projection
- [x] Dedicated place to handle all inits
- [x] Handle deepnorm QKV init
- [x] Handle the small init option
- [x] Typesafe all the inits
- [x] Add a weight init unit test
- [x] Add some explanations in the README
- [ ] Check test coverage
- [x] Check with a bunch of quick trainings whether all the inits are in the green
- [x] Make it possible in the MHA to select the bias per projection 
  
## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
It turns out that using a single projection for QKV in the self-attention case is actually slower, due to the posterior .contiguous() call @dianaml0, I got this from your benchmark_multihead script. Removing this as part of this PR, this simplifies the code a lot in the initialization part as a positive side effect. Having a custom self attention projection which results in contiguous tensors from the get go would fix that, and ideally we could use tensor views to keep the initialization code the same (so that the changes from this PR would not need to be reversed.

Separate projections:
 --- Type: torch.float16 --- 
|  Units: runtime in ms, lower is better               |B=8, M=384, K=128, N_HEADS=4|B=8, M=784, K=512, N_HEADS=4|B=4, M=1024, K=768, N_HEADS=4|B=4, M=2048, K=1024, N_HEADS=4|B=2, M=2048, K=2048, N_HEADS=4|B=2, M=2048, K=4096, N_HEADS=4|B=2, M=4096, K=4096, N_HEADS=4|B=1, M=2048, K=12288, N_HEADS=4|
|------------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
| torch - fw (self_attn)                               |0.24                |1.61                |1.86                |6.59                |7.70                |21.76               |51.08               |73.89               |
| xf - fw (self_attn)                                  |0.19                |1.45                |1.62                |5.71                |7.25                |21.16               |49.29               |75.29               |


With a single QKV projection but a .contiguous() call afterwards:
 --- Type: torch.float16 --- 
|  Units: runtime in ms, lower is better               |B=8, M=384, K=128, N_HEADS=4|B=8, M=784, K=512, N_HEADS=4|B=4, M=1024, K=768, N_HEADS=4|B=4, M=2048, K=1024, N_HEADS=4|B=2, M=2048, K=2048, N_HEADS=4|B=2, M=2048, K=4096, N_HEADS=4|B=2, M=4096, K=4096, N_HEADS=4|B=1, M=2048, K=12288, N_HEADS=4|
|------------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
| torch - fw (self_attn)                               |0.24                |1.62                |1.86                |6.60                |7.75                |21.57               |51.10               |76.29               |
| xf - fw (self_attn)                                  |0.20                |1.47                |1.76                |5.82                |7.42                |21.53               |51.18               |76.19               |

File: xformers/components/multi_head_dispatch.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 from torch.nn.init import constant_
 
 from xformers.components.attention import Attention
-from xformers.components.in_proj_container import InProjContainer, InProjParams
```
Comment (blefaudeux) [inline xformers/components/multi_head_dispatch.py:15]: renaming this, I thought it was a little easier to understand what it was doing from the outside

**Comment by blefaudeux** [comment]
example of a minimal Cifar-10 training

Screenshot from 2022-06-01 11-36-37

File: tests/test_block_factory.py [hunk 2]
```diff
@@ -230,14 +231,14 @@ def test_xformer_decoder_block(
...
 
     encoded = encoder_block(inputs)
-    if decoder_block.mha.attention.supports_attention_mask:
+    if decoder_block.supports_attention_mask:
         _ = decoder_block(
             inputs, encoded, encoder_att_mask=att_mask, input_mask=input_mask
         )
 
     # Test different sequence lengths when encoding and decoding
-    if not decoder_block.mha.attention.requires_same_k_q_dimensions:
```
Comment (blefaudeux) [inline tests/test_block_factory.py:239]: this was getting a little obscure, so the encoder and decoder blocks now expose directly these attention flags, I figured it was easier to read and made sense

File: tests/test_pytorch_transformer_parity.py [hunk 3]
```diff
@@ -172,7 +169,6 @@ def test_pytorch_tranformer_parity(device=torch.device("cuda")):
             dim_feedforward=4 * EMB,
             dropout=DROP,
             activation=ACTIVATION,
-            layer_norm_eps=1e-06,
```
Comment (blefaudeux) [inline tests/test_pytorch_transformer_parity.py:175]: the default is 1e-5 both for xformers/triton and pytorch actually

File: tests/test_pytorch_transformer_parity.py [hunk 4]
```diff
@@ -35,7 +35,6 @@
             "attention": {
                 "name": "scaled_dot_product",
                 "dropout": DROP,
-                "causal": False,
```
Comment (blefaudeux) [inline tests/test_pytorch_transformer_parity.py:38]: this is the default, and less is more

File: xformers/components/multi_head_dispatch.py [hunk 5]
```diff
@@ -59,10 +59,10 @@ def __init__(
         attention: Attention,
         bias: bool = True,
         residual_dropout: float = 0.0,
+        use_separate_proj_weight: bool = True,
         dim_key: Optional[int] = None,
         dim_value: Optional[int] = None,
-        in_proj_container: Optional[InProjContainer] = None,
-        use_separate_proj_weight: Optional[bool] = False,
```
Comment (blefaudeux) [inline xformers/components/multi_head_dispatch.py:65]: this was a nasty default I believe, by default we would use the same weights for q/k/v projections, which for the self attention is probably not a good idea

File: xformers/components/multi_head_dispatch.py [hunk 6]
```diff
@@ -79,33 +79,36 @@ def __init__(
...
+        self.dim_key_head = dim_key // num_heads
+        self.dim_value_head = dim_value // num_heads
         self.dim_model = dim_model
         self.attention = attention
 
         # key, query, value projections for all heads
         # critical options are
         # - are we sharing weights ?
-        # - are we adding biases, and if yes are they shared ?
+        # - are we adding biases ?
```
Comment (blefaudeux) [inline xformers/components/multi_head_dispatch.py:90]: I'm not sure of the amount of nuances to support here, for instance one bias setting per input ? could be a little overkill, but easy to add if need be

File: xformers/factory/block_factory.py [hunk 7]
```diff
@@ -130,12 +130,19 @@ def __init__(self, config: xFormerEncoderConfig, **kwargs):
             residual_scale=residual_scale,
         )
 
-        self.mha = build_multi_head_attention(config.multi_head_config)
-        self.feedforward = build_feedforward(asdict(config.feedforward_config))
+        mha = build_multi_head_attention(config.multi_head_config)
```
Comment (blefaudeux) [inline xformers/factory/block_factory.py:133]: we don't have to keep mha and feedforward as direct members, they were showing up in the model graph and when initializing, making it a little hard to see what was actually in the xformer model.

**Comment by blefaudeux** [comment]
MinGPT training with a couple of options. Not earth shattering, everything works, but there are some measurable differences

Screenshot from 2022-06-01 15-10-40

File: xformers/components/multi_head_dispatch.py [hunk 8]
```diff
@@ -59,10 +59,10 @@ def __init__(
         attention: Attention,
         bias: bool = True,
```
[the remark this answers is not in the document]
Comment (dianaml0) [reply xformers/components/multi_head_dispatch.py:60]: Perhaps would be worth the flexibility?

File: xformers/factory/weight_init.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+import torch
+import torch.nn as nn
+from torch.nn.init import (
+    _calculate_fan_in_and_fan_out,
+    _no_grad_trunc_normal_,
+    _no_grad_uniform_,
+)
+
+
+class xFormerWeightInit(str, Enum):
```
[the remark this answers is not in the document]
Comment (dianaml0) [reply xformers/factory/weight_init.py:26]: Great to have this!

**Review by dianaml0 — APPROVED** [review]
Really nice changes!

File: README.md [hunk 10]
```diff
@@ -196,6 +196,26 @@ Patrick et al., 2021](https://arxiv.org/abs/2106.05392)*
 
 </p></details>
 
+<details><summary>Initializations </summary><p>
+  This is completely optional, and will only occur when generating full models through xFormers, not when picking parts individually.
+  There are basically two initialization mechanisms exposed, but the user is free to initialize weights as he/she sees fit after the fact.
+  - Parts can expose a `init_weights()` method, which define sane defaults
+  - xFormers supports [specific init schemes](xformers/factory/weight_init.py) which *can take precedence* over the init_weights()
```
[the remark this answers is not in the document]
Comment (blefaudeux) [reply README.md:203]: I've checked on https://github.com/facebookresearch/xformers/tree/xformers_init and it's properly rendered, maybe a PR thing..

**Pull request merged**